File: C_Playlist.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <unordered_map>

using namespace std;

int main() {
    int n, k;
    cin >> n >> k;

    vector<pair<int, int>> array(n);
    for (int i = 0; i < n; ++i) {
        int t, b;
        cin >> t >> b;
        array[i] = {b, t};  // store as (b, l)
    }

    sort(array.begin(), array.end());
    pair<int, int> base = array[0];
    array.erase(array.begin());

    // Use a max-heap with priority_queue
    priority_queue<tuple<int, int, int>> heap;
    for (int i = 0; i < n - 1; ++i) {
        int b = array[i].first;
        int l = array[i].second;
        heap.emplace(-l, b, i);
    }

    unordered_map<pair<int, int>, int> hashMap;
    int total = base.second;

    for (int i = 0; i < k - 1 && !heap.empty(); ++i) {
        auto [l, b, _] = heap.top();
        heap.pop();
        total += -l;
        hashMap[{-l, b}]++;
    }

    long long ans = static_cast<long long>(base.first) * total;
    total -= base.second;

    for (int i = 0; i < n - 1; ++i) {
        int b = array[i].first;
        int l = array[i].second;
        if (hashMap[{-l, b}]) {
            hashMap[{-l, b}]--;
            if (hashMap[{-l, b}] == 0) {
                hashMap.erase({-l, b});
            }
        
            while (!heap.empty()) {
                auto [c, d, index] = heap.top();
                heap.pop();
                if (index < i) {
                    continue;
                }
                hashMap[{c, d}]++;
                total += -c;
                break;
            }
        } else {
            total += l;
        }

        ans = max(ans, static_cast<long long>(total) * b);
        total -= l;
    }

    cout << ans << endl;

    return 0;
}

File: C_Playlist.rb
require 'heapq'
require 'set'

n, k = gets.split.map(&:to_i)

array = n.times.map do
  t, b = gets.split.map(&:to_i)
  [b, t]  # Store as [b, l]
end

array.sort!

base = array.shift

# Use a max-heap by negating the length
heap = array.map.with_index { |(b, l), i| [-l, b, i] }
heapq = Heapq.new(heap)

hash_map = Hash.new(0)
total = base[1]

(k - 1).times do
  break if heapq.empty?

  l, b, _ = heapq.heappop
  total += -l
  hash_map[[-l, b]] += 1
end

ans = base[0] * total
total -= base[1]

array.each_with_index do |(b, l), i|
  if hash_map[[-l, b]] > 0
    hash_map[[-l, b]] -= 1
    hash_map.delete([-l, b]) if hash_map[[-l, b]] == 0

    until heapq.empty?
      c, d, index = heapq.heappop
      next if index < i

      hash_map[[c, d]] += 1
      total += -c
      break
    end
  else
    total += l
  end

  ans = [ans, total * b].max
  total -= l
end

puts ans
